### PR TITLE
Swift: add constructor flow

### PR DESCRIPTION
**Pull request by rdmarsh2**
This adds flow from the final value of the `self` object within a constructor to the result of the constructor.

### CODE REVIEW

File: swift/ql/test/library-tests/dataflow/dataflow/DataFlowInline.expected
Comment (MathiasVP) [inline swift/ql/test/library-tests/dataflow/dataflow/DataFlowInline.expected]: It looks like this test is already passing _before_ any dataflow library changes? Are we sure we don't actually have all the required flow already?

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:402]: I'd appreciate a comment on this explaining that it models the `self` value in a constructor / initializer being (implicitly) returned with the newly constructed object.

**Review by geoffw0 — COMMENTED** [review]
This looks great!

We should probably run DCA on the PR before merging.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:402]: Do we really need to create a whole new class for this? Can we not just inherit from `InoutReturnNodeImpl`, restrict the `exit` member already present in that class and override `getKind`?

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:422]: Needs autoformat, I think.

**Review by MathiasVP — COMMENTED** [review]
One comment. Otherwise, I think this looks good (although I hope we can do the simplication I've suggested) once we have a successful DCA run.

**Comment by geoffw0** [comment]
LGTM apart from a merge conflict.  Note that the `swift/ql/test/library-tests/dataflow/taint` test directory was split a few days ago into `swift/ql/test/library-tests/dataflow/taint/core` and `swift/ql/test/library-tests/dataflow/taint/libraries`.  If you merge main and run both of those tests, I think you should be good.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: This has been fixed in https://github.com/github/codeql/pull/12415/commits/86fd21de452bb83850fea9182b3bbd2d78876e7a.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:410]: The refactor you did in https://github.com/github/codeql/pull/12415/commits/86fd21de452bb83850fea9182b3bbd2d78876e7a threw away the constraint that the parameter must be a `SelfParamDecl`. I think we still need this to not have all parameters exit via a `NormalReturnKind`.

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:410]: Good spot.  I'm a little surprised there are no spurious test results though!

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:410]: Yes. It would require a test with a constructor that writes to an inout parameter, which we might not have currently. Something like:
```swift
class X {
  init(_ n : inout Int) { n = source() }
}
func foo() {
  var n = 0
  sink(X(&n))
}
```
i.e., we write to a non-self parameter, but because we're forgetting the `SelfParamDecl` constraint this non-self parameter is leaving the function as the return value.

**Review by MathiasVP — APPROVED** [review]
LGTM once the merge conflict is resolved!

**Comment by MathiasVP** [comment]
Just needs an autoformat now 🤞.

**Review by MathiasVP — APPROVED** [review]

**Comment by geoffw0** [comment]
🎉

**Pull request merged**
